refactor examples/tests + additional `alpaka::exec()` interface

psychocoderHPC opened the pull request:
# new interface
- add new interface `exec(queue, workDiv, kernelBundle)`
- add new interface `createKernelTask(workDiv, kernelBundle)`

We observed a tiny overhead when we start kernel with 100 arguments or an 1000 byte array as argument, see https://github.com/alpaka-group/alpaka/pull/2339#issuecomment-2268504326
Overhead is <1ms for `10000` kernel invocations.

# refactoring
- rename `bundeledKernelFill` into `kernelBundle`
- use the kernel bundle everywhere where we already have this object for the valid work division creation


This PR replaces #2339. Due to performance issues observed in #2339 the `TaskKernel` objects will keep the kernel instances and arguments separately and will not use `KernelBundle`

changes coming from PR https://github.com/alpaka-group/alpaka/pull/2318

mehmetyusufoglu commented:
Great, thanks. I know it is draft, just to add, the only issue is `isKernelArgumentTriviallyCopyable` check. [It was moved into  KernelBundle class at #2318  since it is not used anywhere else and better to check there. Because KernelBundle is used if the kernel is executed somehow; either before exec or inside the overloaded exec.]

mehmetyusufoglu commented on include/alpaka/kernel/Traits.hpp:
```diff
@@ -375,6 +362,30 @@ namespace alpaka
         -> void
     {
-        enqueue(queue, createTaskKernel<TAcc>(workDiv, kernelFnObj, std::forward<TArgs>(args)...));
+        auto const& bundeledKernel = alpaka::KernelBundle(kernelFnObj, std::forward<TArgs>(args)...);
```
typo at variable name `bundeledKernel`, `kernelBundle`

mehmetyusufoglu commented on include/alpaka/kernel/Traits.hpp:
```diff
@@ -328,42 +332,25 @@ namespace alpaka
-//! Executes the given kernel in the given queue.
-//!
+    //! Executes the given kernel in the given queue.
+    //!
```
[nit] additional line not needed.

psychocoderHPC commented:
> Great, thanks. I know it is draft, just to add, the only issue is `isKernelArgumentTriviallyCopyable` check. [It was moved into KernelBundle class at #2318 since it is not used anywhere else and better to check there. Because KernelBundle is used if the kernel is executed somehow; either before exec or inside the overloaded exec.]

Thanks for pointing it out. I found it in #2339 too and reverted it. We can not move this trait away because this will mean it is hard to find for the user. Letzt have a look in a follow-up PR if it is solvable.

mehmetyusufoglu committed:
new interface: support kernel bundle in `exec` and `createTaskKernel`

Co-authored-by: René Widera <r.widera@hzdr.de>

mehmetyusufoglu committed:
refactor examples and tests

- rename `bundeledKernelFill` into `kernelBundle`
- use the kernel bundle everywhere where we already have this object for
  the automatically work division creation

Co-authored-by: René Widera <r.widera@hzdr.de>

psychocoderHPC commented:
@fwyzard Could you have a look into this PR

fwyzard commented on example/heatEquation/src/heatEquation.cpp:
```diff
@@ -146,8 +146,10 @@ auto example(TAccTag const&) -> int
+
         // Compute next values
-        alpaka::exec<Acc>(queue, workDiv, heatEqKernel, pCurrAcc, pNextAcc, numNodesX, dx, dt);
+        alpaka::exec<Acc>(queue, workDiv, tmpKernelBundle);
```
This is one example where I would prefer _not to_ introduce the extra kernel bundle, since it makes the code longer.

psychocoderHPC replied:
It uses `getValidWorkDivForKernel` therefore the bundle is required anyway and it is already in the mainline. This changing the kernel call to avoid calling the kernel with different arguments as used in `getValidWorkDivForKernel`

fwyzard replied:
It uses `getValidWorkDivForKernel` with one bundle, but it does not use _that_ bundle to launch the kernel, it creates a new one _only_ to launch the kernel with different arguments.

There is no reason to create and use the second bundle.

This is exactly the use case where one should _not_ use a bundle, and call `exec `directly with the arguments.

fwyzard commented on example/helloWorld/src/helloWorld.cpp:
```diff
@@ -133,12 +133,11 @@ auto example(TAccTag const&) -> int
     // callable operator() and takes the accelerator as first
     // argument. So a kernel can be a class or struct, a lambda, etc.
-    HelloWorldKernel helloWorldKernel;
+    alpaka::KernelBundle kernelBundle = HelloWorldKernel{};
```
Shouldn't this be
```c++
alpaka::KernelBundle kernelBundle = alpaka::KernelBundle(HelloWorldKernel{});
```
?

fwyzard replied:
Ah, the explicit constructor call is not needed, because it is not marked as `explicit` ?

psychocoderHPC replied:
`auto kernelBundle = alpaka::KernelBundle(HelloWorldKernel{});`

will be the long why to write the same line of code I did. The constructor is no marked `explicit` but all types will be deducted because the second parameter of the constructor is a variadic template.

fwyzard commented on example/randomCells2D/src/randomCells2D.cpp:
```diff
@@ -202,16 +202,17 @@ auto example(TAccTag const&) -> int
 
-    alpaka::exec<Acc>(queue, workDivInitRandom, initRandomKernel, extent, ptrBufAccRandV, pitchBufAccRandV);
+    // execute the same kernel with different pointers
+    alpaka::exec<Acc>(queue, workDivInitRandom, kernelBundleInitRandom);
```
Isn't the previous call the one that should be replace with the `kernelBundle` ?
This one uses different arguments.

psychocoderHPC replied:
well catched, thats a bug I introduced.

fwyzard commented on example/randomStrategies/src/randomStrategies.cpp:
```diff
@@ -266,11 +266,7 @@ void runStrategy(Box<TAccTag>& box)
-        ptrBufAccRand,
-        static_cast<unsigned>(
-            box.extentResult[0] / box.extentRand[0])); // == NUM_ROLLS; amount of work to be performed by each thread
+        kernelBundle); // == NUM_ROLLS; amount of work to be performed by each thread
```
I think the comment should be moved to line 254.

psychocoderHPC replied:
I will remove this comment here. We have this comment on the line where `NUM_ROLLS` is used, I do not know why the comment is there. The coresponding lines have nothing to do with `NUM_ROLLS`

fwyzard commented on include/alpaka/kernel/KernelBundle.hpp:
```diff
@@ -32,7 +32,6 @@ namespace alpaka
         {
         }
 
-    private:
```
Why no longer `private` ?

psychocoderHPC replied:
We access the member from outside. Writing getter would be possible but IMO not strictly required

psychocoderHPC commented:
@fwyzard I updated the PR.

fwyzard commented on test/integ/sharedMem/src/sharedMem.cpp:
```diff
@@ -160,8 +161,11 @@ TEMPLATE_LIST_TEST_CASE("sharedMem", "[sharedMem]", TestAccs)
-    auto const taskKernel = alpaka::createTaskKernel<Acc>(workDiv, kernel, std::data(blockRetValsAcc));
+
+    // Create the kernel execution task using the real data.
+    auto const& kernelBundle2 = alpaka::KernelBundle(kernel, std::data(blockRetValsAcc));
```
I think this could be simply
```suggestion
    auto const kernelBundle2 = alpaka::KernelBundle(kernel, std::data(blockRetValsAcc));
```
since it's a constructor call ?

fwyzard replied:
However, I don't like that the bundle is created only to execute the kernel.

This is another example where I don't like forcing the use of the bundles.

psychocoderHPC replied:
But it is required for `getValidWorkDivKernel`. We kept therefore the old interface but if you use `getValidWorkDivKernel` you should use the bundle for the kernel call to avoid signature mismatch which can end with a kernel crash due to invalid work extents.

fwyzard replied:
I understand - but we are not using `kernelBundle2` with `getValidWorkDivKernel`.

psychocoderHPC replied:
fair point. In prinziple the example is broken because the workdiv is assumed to work for the second kernel too. I will change this line back

fwyzard commented on test/integ/separableCompilation/src/main.cpp:
```diff
@@ -111,23 +111,21 @@ TEMPLATE_LIST_TEST_CASE("separableCompilation", "[separableCompilation]", TestAc
-        = alpaka::KernelBundle(kernel, memBufAccA.data(), memBufAccB.data(), memBufAccC.data(), numElements);
+    auto const& kernelBundle = alpaka::KernelBundle(
+        kernel,
+        std::data(memBufAccA),
```
I do like `x.data()` more than `std::data(x)` (it's shorter, and the latter does not gain anything), but I guess it's a matter of taste.

psychocoderHPC replied:
yes this is matter of tast and was in the code before too.

fwyzard commented on include/alpaka/test/KernelExecutionFixture.hpp:
```diff
@@ -69,21 +69,22 @@ namespace alpaka::test
+            auto const& kernelBundle
+                = alpaka::KernelBundle<TKernelFnObj, decltype(getPtrNative(bufAccResult)), TArgs...>(
+                    kernelFnObj,
+                    getPtrNative(bufAccResult),
```
```suggestion
                    bufAccResult.data(),
```
or
```suggestion
                    std::data(bufAccResult),
```
?

psychocoderHPC replied:
I moved only code without changing it.

fwyzard commented on example/randomCells2D/src/randomCells2D.cpp:
```diff
@@ -230,7 +231,7 @@ auto example(TAccTag const&) -> int
     RunTimestepKernelSingle runTimestepKernelSingle;
 
-    auto const& bundeledKernelRuntimeStep = alpaka::KernelBundle(
+    auto const& kernelBundleRuntimeStep = alpaka::KernelBundle(
```
```suggestion
    auto const kernelBundleRuntimeStep = alpaka::KernelBundle(
```
?

psychocoderHPC replied:
We have it already in the mainline. It is valid because `const&` is extending the live time of the temporary. 
I am no fan of using it but since @mehmetyusufoglu did it already everywhere I kept it.

fwyzard replied:
I see... OK, we can fix it later :-)

fwyzard commented on example/randomStrategies/src/randomStrategies.cpp:
```diff
@@ -247,7 +247,7 @@ void runStrategy(Box<TAccTag>& box)
 
 
-    auto const& bundeledKernel = alpaka::KernelBundle(
+    auto const& kernelBundle = alpaka::KernelBundle(
```
```suggestion
    auto const kernelBundle = alpaka::KernelBundle(
```

fwyzard commented on example/randomStrategies/src/randomStrategies.cpp:
```diff
@@ -291,27 +284,20 @@ void runStrategy(Box<TAccTag>& box)
 
-    auto const& bundeledKernelFill
-        = alpaka::KernelBundle(fillKernel, box.extentResult, ptrBufAccRand, ptrBufAccResult);
+    auto const& kernelBundleFill = alpaka::KernelBundle(fillKernel, box.extentResult, ptrBufAccRand, ptrBufAccResult);
```
```suggestion
    auto const kernelBundleFill = alpaka::KernelBundle(fillKernel, box.extentResult, ptrBufAccRand, ptrBufAccResult);
```

fwyzard commented on example/convolution1D/src/convolution1D.cpp:
```diff
@@ -140,7 +140,7 @@ auto example(TAccTag const&) -> int
     DataType* nativeOutputDeviceMemory = std::data(outputDeviceMemory);
 
-    auto const& bundeledKernel = alpaka::KernelBundle(
+    auto const& kernelBundle = alpaka::KernelBundle(
```
```suggestion
    auto const kernelBundle = alpaka::KernelBundle(
```

fwyzard commented:
hi @psychocoderHPC, I've updated my benchmark to compare the performance of `alpaka::exec` with a kernel with explicit arguments, and the performance of `alpaka::exec` with the corresponding kernel bundle: 
[benchmark_2340.tar.gz](https://github.com/user-attachments/files/16530776/benchmark_2340.tar.gz)

The `Makefile` is simpler since this PR does not change the behaviour of `alpaka::exec` :-)

Running the two benchmarks on my laptop, this time pinning all jobs to a single core, I get:

### kernel.cc
```bash
$ taskset -c 15 ./kernel.cpu >& /dev/null; for N in {0..99}; do taskset -c 15 ./kernel.cpu | tail -n1; done
...
10000 launches took KernelOneArg:   1.16 ms     KernelStruct:   2.07 ms KernelArray:  12.67 ms
10000 launches took KernelOneArg:   1.26 ms     KernelStruct:   2.06 ms KernelArray:  12.92 ms
10000 launches took KernelOneArg:   1.14 ms     KernelStruct:   2.06 ms KernelArray:  11.32 ms
```

### bundle.cc
```bash
$ taskset -c 15 ./bundle.cpu >& /dev/null; for N in {0..99}; do taskset -c 15 ./bundle.cpu | tail -n1; done
...
10000 launches took KernelOneArg:   1.17 ms     KernelStruct:   3.30 ms KernelArray:  13.10 ms
10000 launches took KernelOneArg:   1.17 ms     KernelStruct:   3.30 ms KernelArray:  13.24 ms
10000 launches took KernelOneArg:   1.21 ms     KernelStruct:   2.10 ms KernelArray:  14.93 ms
```

So, using a kernel bundles is slower by a small but consistent amount.

I will try to understand where the slowdown comes from - but that can be fixed in a follow-up PR.

fwyzard commented:
By the way, I noticed another issue with the `alpaka::KernelBundle`: it is possible to construct a `KernelBundle` object with a kernel and a set or arguments that are incompatible with that kernel.

The compiler will complain only when one tries to use the bundle, but not when it is constructed.

psychocoderHPC commented:
> By the way, I noticed another issue with the `alpaka::KernelBundle`: it is possible to construct a `KernelBundle` object with a kernel and a set or arguments that are incompatible with that kernel.
> 
> The compiler will complain only when one tries to use the bundle, but not when it is constructed.

This can not avoided because we do not have the accelerator at this point to do the signature verification. This is done in `TaskKernel`

fwyzard commented:
> > By the way, I noticed another issue with the `alpaka::KernelBundle`: it is possible to construct a `KernelBundle` object with a kernel and a set or arguments that are incompatible with that kernel.
> > The compiler will complain only when one tries to use the bundle, but not when it is constructed.
> 
> This can not avoided because we do not have the accelerator at this point to do the signature verification. This is done in `TaskKernel`

I see... and we need the accelerator because the kernel is templated on the accelerator type 🤔

psychocoderHPC commented:
> > > By the way, I noticed another issue with the `alpaka::KernelBundle`: it is possible to construct a `KernelBundle` object with a kernel and a set or arguments that are incompatible with that kernel.
> > > The compiler will complain only when one tries to use the bundle, but not when it is constructed.
> > 
> > 
> > This can not avoided because we do not have the accelerator at this point to do the signature verification. This is done in `TaskKernel`
> 
> I see... and we need the accelerator because the kernel is templated on the accelerator type 🤔

Yes that's the problem. In the first drafts we carried the bundle with the accelerator but in general, this is also not nice if you would like to re-use the kernel for a different accelerator.

psychocoderHPC commented:
@fwyzard Is there a real show stopper for merging this code. IMO your comments are mostly style or best practices.
Let us talk about this in a developer meeting and discuss how we would like to handle `.data()`, `getNativePtr()` and `auto const & = KernelBundle` and apply changes in an orthogonal PR to the full code base.

fwyzard commented:
I would like to revert these two changes, where the kernel bundle is used unnecessarily:

### `example/heatEquation/src/heatEquation.cpp`
```diff
+        auto const& tmpKernelBundle = alpaka::KernelBundle(heatEqKernel, pCurrAcc, pNextAcc, numNodesX, dx, dt);
+
         // Compute next values
-        alpaka::exec<Acc>(queue, workDiv, heatEqKernel, pCurrAcc, pNextAcc, numNodesX, dx, dt);
+        alpaka::exec<Acc>(queue, workDiv, tmpKernelBundle);
```

### `test/integ/sharedMem/src/sharedMem.cpp`
```diff
-    // Create the kernel execution task.
-    auto const taskKernel = alpaka::createTaskKernel<Acc>(workDiv, kernel, std::data(blockRetValsAcc));
+
+    // Create the kernel execution task using the real data.
+    auto const& kernelBundle2 = alpaka::KernelBundle(kernel, std::data(blockRetValsAcc));
+    auto const taskKernel = alpaka::createTaskKernel<Acc>(workDiv, kernelBundle2);
+
```

The rest is OK.

psychocoderHPC commented:
> I would like to revert these two changes, where the kernel bundle is used unnecessarily:
> 
> ### `example/heatEquation/src/heatEquation.cpp`
> ```diff
> +        auto const& tmpKernelBundle = alpaka::KernelBundle(heatEqKernel, pCurrAcc, pNextAcc, numNodesX, dx, dt);
> +
>          // Compute next values
> -        alpaka::exec<Acc>(queue, workDiv, heatEqKernel, pCurrAcc, pNextAcc, numNodesX, dx, dt);
> +        alpaka::exec<Acc>(queue, workDiv, tmpKernelBundle);
> ```
> 
> ### `test/integ/sharedMem/src/sharedMem.cpp`
> ```diff
> -    // Create the kernel execution task.
> -    auto const taskKernel = alpaka::createTaskKernel<Acc>(workDiv, kernel, std::data(blockRetValsAcc));
> +
> +    // Create the kernel execution task using the real data.
> +    auto const& kernelBundle2 = alpaka::KernelBundle(kernel, std::data(blockRetValsAcc));
> +    auto const taskKernel = alpaka::createTaskKernel<Acc>(workDiv, kernelBundle2);
> +
> ```
> 
> The rest is OK.

done

fwyzard commented:
By the way, after reading the code for `alpaka::exec`, `alpaka::createTaskKernel` and `alpaka::KernelBundle`, I think that the extra copy of the kernel arguments is unavoidable when using the bundle.

With the current implementation, constructing the `KernelBundle` requires copying the arguments into the bundle itself.

Changing the bundle to store only a `const&` to the argument would avoid the copy, but it would not be safe: the arguments could go out of scope, or be modified, after the bundle is created but before the kernel is launched.

Changing the bundle to "move" the arguments when launching the kernel may not be feasible, may not be safe (if we reuse the bundle for multiple launches), it wouldn't help anyway for types like large structs.

One possible (future) improvement could be to change the way the CUDA kernels are launched to use `cudaLaunchKernel` and use the bundle to store the temporary arguments, instead of making an extra copy. However, this helps only for the CUDA backend (where the launch cost is anyway much larger that the cost of the extra copy), not for the various CPU backends.

psychocoderHPC commented:
@fwyzard I have set this PR to draft that it not get merged. There is an alternative to the kernel bundle. We could also change the `getValidWorkDivKernel` function signature to avoid the kernel bundle completly. 
This is before the release still possible.

fwyzard commented:
I'm actually fine to merge the PR, as long as the alternative of not using the `KernelBundle` is kept.

On the other hand, we could have a look at what a kernel graph node looks like in CUDA graphs, and see if can adapt the `KernalBundle` to match ?

psychocoderHPC commented:
this PR is closed because we switched back to old interfaces where kernel bundle is not used anymore https://github.com/alpaka-group/alpaka/pull/2349
